pkg/crypto/tinkcrypto/primitive/aead/subtle/aes_cbc.go

```diff
@@ -0,0 +1,154 @@
+	}
+
+	if n := copy(ciphertext[AESCBCIVSize:], plaintext); n != plainTextSize {
+		return nil, fmt.Errorf("aes_cbc: failed to copy plaintext (copied %d/%d bytes)", n, plainTextSize)
+	}
+
+	ciphertext = Pad(ciphertext, plainTextSize, cbc.BlockSize())
+
+	cbc.CryptBlocks(ciphertext[AESCBCIVSize:], ciphertext[AESCBCIVSize:])
+
+	return ciphertext, nil
+}
+
+// Decrypt decrypts ciphertext.
+func (a *AESCBC) Decrypt(ciphertext []byte) ([]byte, error) {
+	ciphertextSize := len(ciphertext)
+	if ciphertextSize < AESCBCIVSize {
+		return nil, errors.New("aes_cbc: ciphertext too short")
+	}
+
+	iv := ciphertext[:AESCBCIVSize]
+
+	cbc, err := newCipher(a.Key, iv, true)
+	if err != nil {
+		return nil, fmt.Errorf("aes_cbc: Decrypt() %w", err)
+	}
+
+	blockSize := cbc.BlockSize()
+
+	if len(ciphertext[AESCBCIVSize:])%blockSize > 0 {
+		return nil, errors.New("aes_cbc: invalid ciphertext padding")
+	}
+
+	plaintext := make([]byte, ciphertextSize-AESCBCIVSize)
+	cbc.CryptBlocks(plaintext, ciphertext[AESCBCIVSize:])
+
+	if len(plaintext) == 0 {
+		return plaintext, nil
+	}
+
+	// unpad plaintext if not block size.
+	last := plaintext[len(plaintext)-1]
+	count := int(last)
+
+	if count == 0 || count > blockSize || count > len(plaintext) {
+		return nil, errors.New("aes_cbc: invalid padding")
+	}
+
+	padding := bytes.Repeat([]byte{last}, count)
+	if !bytes.HasSuffix(plaintext, padding) { // padding not found, return full plaintext.
+		return plaintext, nil
+	}
+
+	// padding was found, trim it and return remaining plaintext.
+	return plaintext[:len(plaintext)-len(padding)], nil
+}
+
+// newIV creates a new IV for encryption.
+func (a *AESCBC) newIV() []byte {
+	return random.GetRandomBytes(uint32(AESCBCIVSize))
+}
+
+// newCipher creates a new AES-CBC cipher using the given key, IV and the crypto library.
+func newCipher(key, iv []byte, decrypt bool) (cipher.BlockMode, error) {
+	block, err := aes.NewCipher(key)
+	if err != nil {
+		return nil, fmt.Errorf("aes_cbc: failed to create block cipher, error: %w", err)
+	}
+
+	// If the IV is less than BlockSize bytes we need to pad it with zeros otherwise NewCBCEncrypter will panic.
+	if len(iv) < aes.BlockSize {
+		paddedIV := make([]byte, aes.BlockSize)
+		if n := copy(paddedIV, iv); n != len(iv) {
+			return nil, errors.New("aes_cbc: failed to pad IV")
+		}
+
+		if !decrypt {
+			return cipher.NewCBCEncrypter(block, paddedIV), nil
+		}
+
+		return cipher.NewCBCDecrypter(block, paddedIV), nil
+	}
+
+	if !decrypt {
+		return cipher.NewCBCEncrypter(block, iv), nil
+	}
+
+	return cipher.NewCBCDecrypter(block, iv), nil
+}
+
+// Pad text to blockSize.
+func Pad(text []byte, originalTextSize, blockSize int) []byte {
+	// pad to block size if needed.
+	missing := blockSize - (originalTextSize % blockSize)
+	if missing > 0 && originalTextSize > 0 {
+		text = append(text, bytes.Repeat([]byte{byte(missing)}, missing)...)
+	}
+
+	return text
+}
```

Comment:
Don't make padding optional because otherwise the ciphertext of the empty plaintext would be distinguishable. This violates semantic security.

Add a block of padding when the plaintext size is a multiple of the blocksize.

Reply:
Hmm I'm wrong because IIRC semantic security ignores message length. FWIW, CTF also leaks when the empty plaintext is encrypted.

I still think you should add a padding block. Page 17 of https://nvlpubs.nist.gov/nistpubs/Legacy/SP/nistspecialpublication800-38a.pdf also recommends that.

Reply:
I see, thank you for this comment, I'll force add an extra padding block as per Page 17 of this NIST doc, quoted below:

> One way to ensure that the receiver does not mistakenly remove bits from an unpadded message is to require the sender to pad every message, including messages in which the final block (segment) is already complete. For such messages, an entire block (segment) of padding is appended.